### PR TITLE
Upgrade HyperFormula to 1.1.0

### PR DESCRIPTION
**Pull request by nikersify**
### Context
This PR updates the HyperFormula dependency to `1.1.0`.

The following tests were removed due to changes in how hf handles matrices (now called arrays):
- `should block removing (rows|columns) if there\'s a matrix underneath`
- `should not autofill if there\'s a matrix in the way`
- `should block removing rows based on physical indexes`
- `should block removing columns based on physical indexes`
- `should block adding rows based on physical indexes`
- `should block adding columns based on physical indexes`

The custom autofilling logic was removed in favor of new hf's built in functionality.

A potential todo would be a migration guide with all the things that have changed since `hf@0.x`.

### How has this been tested?
With the existing test suite.

### Related issue(s):
1. #8502

### Affected project(s):
- [x] `handsontable`

### CODE REVIEW

File: src/plugins/formulas/formulas.js [hunk 1]
```diff
@@ -32,6 +31,20 @@ Hooks.getSingleton().register('afterSheetRemoved');
...
+ * @param {Function} fn Function to surround with try-catch.
+ * @returns {*}
+ */
+const attempt = (fn) => {
+  try {
+    return fn();
+  } catch (_) {
+    // eslint-disable-next-line no-empty
+  }
+};
```
Comment (wojciechczerniak) [inline src/plugins/formulas/formulas.js:46]: Why is this wrapper function necessary?

File: src/plugins/formulas/__tests__/publicAPI.spec.js [hunk 2]
```diff
@@ -72,15 +72,15 @@ describe('Formulas public API', () => {
...
 
       const formulas = getPlugin('formulas');
 
       expect(formulas.getCellType(0, 0)).toBe('VALUE');
       expect(formulas.getCellType(0, 1)).toBe('VALUE');
       expect(formulas.getCellType(1, 0)).toBe('VALUE');
-      expect(formulas.getCellType(2, 0)).toBe('MATRIX');
-      expect(formulas.getCellType(3, 0)).toBe('MATRIX');
+      expect(formulas.getCellType(2, 0)).toBe('ARRAYFORMULA');
+      expect(formulas.getCellType(3, 0)).toBe('VALUE');
```
Comment (wojciechczerniak) [inline src/plugins/formulas/__tests__/publicAPI.spec.js:83]: Isn't this going to be ARRAY type? The default configuration has `useArrayArithmetic` disabled. But then, we need to either change `=TRANSPOSE(A1:B2)` to `=ARRAYFORMULA(A2:A5*B2:B5)` that will result in an array or add another test where `useArrayArithmetic` is enabled. ATM we have just one edge case

```suggestion
      expect(formulas.getCellType(3, 0)).toBe('ARRAY');
```

File: src/plugins/formulas/engine/settings.js [hunk 3]
```diff
@@ -5,9 +5,6 @@ export const DEFAULT_LICENSE_KEY = 'internal-use-in-handsontable';
 const DEFAULT_SETTINGS = {
   licenseKey: DEFAULT_LICENSE_KEY,
 
-  binarySearchThreshold: 20,
-  matrixDetection: false,
-  matrixDetectionThreshold: 100,
   useColumnIndex: false,
```
Comment (wojciechczerniak) [inline src/plugins/formulas/engine/settings.js:8]: All removed options are correct, but we have to add new option as well
```suggestion
  useArrayArithmetic: false,
  useColumnIndex: false,
```

File: src/plugins/formulas/formulas.js [hunk 4]
```diff
@@ -384,7 +398,7 @@ export class Formulas extends BasePlugin {
   isFormulaCellType(row, column, sheet = this.sheetId) {
     const cellType = this.getCellType(row, column, sheet);
 
-    return cellType === 'FORMULA' || cellType === 'MATRIX';
+    return cellType === 'FORMULA' || cellType === 'ARRAYFORMULA';
   }
```
Comment (wojciechczerniak) [inline src/plugins/formulas/formulas.js:402]: It does not make sense to fix this method each time something changes in HF and HF has exactly the same method already: https://github.com/handsontable/hyperformula/blob/b303691371d5ab9a66f5eae050cdfbf596530613/src/HyperFormula.ts#L3096 can we just use HF instead?

```ts
isFormulaCellType(row, column, sheet = this.sheetId) {
  return this.engine.doesCellHaveFormula({ row, column, sheet });
}
```

File: src/plugins/formulas/formulas.js (hunk 1 above)
Comment (wojciechczerniak) [reply src/plugins/formulas/formulas.js:46]: It will suppress error msg from HF that the cellAddress is invalid. Users should know why it does not work, right?

File: .changelogs/8502.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+  "title": "Updated HyperFormula to 1.1.0",
```
Comment (wojciechczerniak) [inline .changelogs/8502.json:2]: More like `^1.x.x`. And a short explanation why this is a breaking change. Maybe @kirszenbaum can help how to fit it in a single sentence? We will probably go into details in a blog post, but should be more informative at a glance. i.e. if someone is not using formulas will know that "Breaking change" is not for them

File: src/plugins/formulas/formulas.js (hunk 4 above)
Comment (jansiegel) [reply src/plugins/formulas/formulas.js:402]: +1, currently `doesCellHaveFormula` has the exact same condition built-in: https://github.com/handsontable/hyperformula/blob/072f829bff49a389b2b4be8649cf654bdb998cf2/src/HyperFormula.ts#L3096 (in `<1.0.0` the right-hand side of the condition was missing there)

File: src/plugins/formulas/formulas.js [hunk 6]
```diff
@@ -522,6 +536,45 @@ export class Formulas extends BasePlugin {
...
+    };
+    const source = { start: sourceLeftCorner, end: sourceEnd };
+    const targetEnd = {
+      row: targetLeftCorner.row + targetHeight - 1,
+      col: targetLeftCorner.col + targetWidth - 1,
+      sheet: targetLeftCorner.sheet,
+    };
+    const target = { start: targetLeftCorner, end: targetEnd };
+
+    return this.engine.getFillRangeData(source, target);
```
Comment (jansiegel) [inline src/plugins/formulas/formulas.js:575]: 1. The `isItPossibleToSetCellContents` currently only takes 1 argument (a single cell address or a cell range): https://handsontable.github.io/hyperformula/api/classes/hyperformula.html#isitpossibletosetcellcontents 
2. I think the code may be simplified a little bit, as the `CellRange` class contains methods like `getTopLeftCorner` and `getBottomRightCorner`, so unless I'm missing something, there's no need for retrieving the width/height of the ranges and manually calculating the ends of ranges.

Maybe something like this could also work?

```suggestion
    const extendWithSheetId = range => Object.assign(range, { sheet: this.sheetId });
    const engineSourceRange = {
      start: extendWithSheetId(sourceRange.getTopLeftCorner()),
      end: extendWithSheetId(sourceRange.getBottomRightCorner())
    };
    const engineTargetRange = {
      start: extendWithSheetId(targetRange.getTopLeftCorner()),
      end: extendWithSheetId(targetRange.getBottomRightCorner())
    };

    // Blocks the autofill operation if HyperFormula says that at least one of
    // the underlying cell's contents cannot be set.
    if (this.engine.isItPossibleToSetCellContents(engineTargetRange) === false) {
      return false;
    }

    return this.engine.getFillRangeData(engineSourceRange, engineTargetRange);
```

**Review by jansiegel — CHANGES_REQUESTED** [review]

File: .changelogs/8502.json (hunk 5 above)
Comment (kirszenbaum) [reply .changelogs/8502.json:2]:
```suggestion
  "title": "Updated the optional HyperFormula dependency from 0.6.2 to ^1.0.0, which introduces breaking changes for the `formula` plugin users.",
```

File: src/plugins/formulas/__tests__/publicAPI.spec.js (hunk 2 above)
Comment (nikersify) [reply src/plugins/formulas/__tests__/publicAPI.spec.js:83]: Arrays don't seem to work that well at the moment (this spec produces a `SPILL` error), I only made this test pass by changing the spec definition. Matrices were broken in the past as well.

File: src/plugins/formulas/engine/settings.js (hunk 3 above)
Comment (nikersify) [reply src/plugins/formulas/engine/settings.js:9]: https://github.com/handsontable/handsontable/pull/8669/commits/aadcbf01170905b230c8ab35424b574781e7aad1

File: src/plugins/formulas/formulas.js (hunk 1 above)
Comment (nikersify) [reply src/plugins/formulas/formulas.js:46]: Yea I agree it's not pretty, after bumping hf I discovered that we were often calling `hf#getCellType`, `hf#isItPossibleToAdd(Rows|Columns)` and `hf#off` with invalid parameters, mostly wrong indexes. This should be covered by `Rethink renderable/logical/physical indexes support` in https://github.com/handsontable/handsontable/issues/8511.

File: src/plugins/formulas/formulas.js (hunk 4 above)
Comment (nikersify) [reply src/plugins/formulas/formulas.js]: https://github.com/handsontable/handsontable/pull/8669/commits/03704761f213f093b0c63e46d5b9297c562dace5

File: src/plugins/formulas/formulas.js (hunk 1 above)
Comment (nikersify) [reply src/plugins/formulas/formulas.js:46]: Specifically, hot itself returns `null` when calling `toPhysical(Row|Column)` with an out of bounds index.

Another dirty fix with one less `attempt` call, all the tests pass but there might be edge cases: https://github.com/handsontable/handsontable/pull/8669/commits/29fcdfb47b20decf8004f931f5a414fc378f1b4f

File: src/plugins/formulas/formulas.js [hunk 7]
```diff
@@ -594,7 +647,7 @@ export class Formulas extends BasePlugin {
     }
 
     // `column` is here as visual index because of inconsistencies related to hook execution in `src/dataMap`.
-    const isFormulaCellType = this.isFormulaCellType(this.hot.toVisualRow(row), column);
+    const isFormulaCellType = attempt(() => this.isFormulaCellType(this.hot.toVisualRow(row), column));
```
[the remark this answers is not in the document]
Comment (nikersify) [reply src/plugins/formulas/formulas.js]: https://github.com/handsontable/handsontable/pull/8669/commits/1ca1889dfaa5e1a14e245a27853da6ae1178caf7, akin to https://github.com/handsontable/handsontable/pull/8669#discussion_r699517730

File: src/plugins/formulas/formulas.js (hunk 6 above)
Comment (nikersify) [reply src/plugins/formulas/formulas.js]: Seems to work great, thanks 👍 https://github.com/handsontable/handsontable/pull/8669/commits/fb6bd9eb0b257aa22a2f646d79c6a375ede19868

File: src/plugins/formulas/formulas.js [hunk 8]
```diff
@@ -363,13 +377,13 @@ export class Formulas extends BasePlugin {
...
-   * @returns {string} Possible values: 'FORMULA' | 'VALUE' | 'MATRIX' | 'EMPTY'.
+   * @returns {string} Possible values: 'FORMULA' | 'VALUE' | 'ARRAYFORMULA' | 'EMPTY'.
    */
   getCellType(row, column, sheet = this.sheetId) {
     return this.engine.getCellType({
       sheet,
-      row: this.hot.toPhysicalRow(row),
-      col: this.hot.toPhysicalColumn(column)
+      row: this.hot.toPhysicalRow(row) ?? Infinity,
+      col: this.hot.toPhysicalColumn(column) ?? Infinity
```
Comment (wojciechczerniak) [inline src/plugins/formulas/formulas.js:386]: IMO this is a bad idea. We're reading `getCellType` from different cell that may be out of bounds

File: src/plugins/formulas/formulas.js [hunk 9]
```diff
@@ -382,9 +396,11 @@ export class Formulas extends BasePlugin {
    * @returns {boolean}
    */
   isFormulaCellType(row, column, sheet = this.sheetId) {
-    const cellType = this.getCellType(row, column, sheet);
-
-    return cellType === 'FORMULA' || cellType === 'MATRIX';
+    return this.engine.doesCellHaveFormula({
+      sheet,
+      row: this.hot.toPhysicalRow(row) ?? Infinity,
+      col: this.hot.toPhysicalColumn(column) ?? Infinity
```
Comment (wojciechczerniak) [inline src/plugins/formulas/formulas.js:402]: Same here

File: src/plugins/formulas/formulas.js [hunk 10]
```diff
@@ -762,7 +810,12 @@ export class Formulas extends BasePlugin {
    * @returns {*|boolean} If false is returned the action is canceled.
    */
   onBeforeCreateRow(row, amount) {
-    if (!this.engine.isItPossibleToAddRows(this.sheetId, [this.toPhysicalRowPosition(row), amount])) {
+    // Using `attempt` because this.sheetId can be null or point to an non-existing sheet
```
Comment (wojciechczerniak) [inline src/plugins/formulas/formulas.js:813]: So we should check it correctly. Not suppress the error
```suggestion
    if (this.sheetId && this.engine.doesSheetExist(this.sheetId)) {
```

File: src/plugins/formulas/formulas.js [hunk 11]
```diff
@@ -776,7 +829,12 @@ export class Formulas extends BasePlugin {
    * @returns {*|boolean} If false is returned the action is canceled.
    */
   onBeforeCreateCol(col, amount) {
-    if (!this.engine.isItPossibleToAddColumns(this.sheetId, [this.toPhysicalColumnPosition(col), amount])) {
+    // Using `attempt` because this.sheetId can be null or point to an non-existing sheet
```
Comment (wojciechczerniak) [inline src/plugins/formulas/formulas.js:832]: Same here. Check the conditions or make sure that `this.sheetId` always exist. It should

File: src/plugins/formulas/__tests__/formulas.spec.js [hunk 12]
```diff
@@ -2356,4 +2218,28 @@ describe('Formulas general', () => {
...
+        [1, 2, 3, 4, 5],
+        [9, 8, 7, 6, '=myOwnCalc'],
+      ],
+      formulas: {
+        engine: HyperFormula,
+        sheetName: 'my-sheet',
+        namedExpressions: [
+          {
+            name: 'myOwnCalc',
+            expression: '=\'my-sheet\'!$A$1+100',
```
Comment (wojciechczerniak) [inline src/plugins/formulas/__tests__/formulas.spec.js:2234]: That's not the case from #8057. Quoted sheet names are correctly handled and tested by HF. We were discussion better handling misuse scenario:

```suggestion
            expression: '=my-sheet!$A$1+100',
```

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]

File: src/plugins/formulas/formulas.js (hunk 1 above)
Comment (wojciechczerniak) [reply src/plugins/formulas/formulas.js:46]: When hot returns `null` from `toPhysical*` we should handle that internally and don't call HF.  Same thing when it returns something that is out of bounds of HF maxCols and maxRows. But as long as we keep them HF and HOT in sync on those we can skip it.

"Rethink indexes" does not mean we can use invalid Infinity values now.

Comment (jansiegel) [reply src/plugins/formulas/formulas.js]: I removed the `attempt` function in https://github.com/handsontable/handsontable/pull/8669/commits/c52a4f85f0966f7c2eeed8e267a374aff3480695

File: src/plugins/formulas/__tests__/publicAPI.spec.js (hunk 2 above)
Comment (jansiegel) [reply src/plugins/formulas/__tests__/publicAPI.spec.js]: This seems to have been caused by array-based functions' results not being correctly transferred to handsontable, I think it should work fine after https://github.com/handsontable/handsontable/pull/8669/commits/c52a4f85f0966f7c2eeed8e267a374aff3480695. (also, I modified the test case to use both `TRANSPOSE` and `ARRAYFORMULA`)

File: src/plugins/formulas/formulas.js (hunk 8 above)
Comment (jansiegel) [reply src/plugins/formulas/formulas.js]: The `Infinity` mentions should be removed after https://github.com/handsontable/handsontable/pull/8669/commits/d96ae619e5822d099e6120b57403f4785d3c831f.

File: src/plugins/formulas/formulas.js (hunk 9 above)
Comment (jansiegel) [reply src/plugins/formulas/formulas.js]: The `Infinity` mentions should be removed after https://github.com/handsontable/handsontable/pull/8669/commits/d96ae619e5822d099e6120b57403f4785d3c831f.

File: src/plugins/formulas/formulas.js (hunk 10 above)
Comment (jansiegel) [reply src/plugins/formulas/formulas.js]: I extended these conditions in https://github.com/handsontable/handsontable/pull/8669/commits/c52a4f85f0966f7c2eeed8e267a374aff3480695.

File: src/plugins/formulas/formulas.js (hunk 11 above)
Comment (jansiegel) [reply src/plugins/formulas/formulas.js]: I extended these conditions in https://github.com/handsontable/handsontable/pull/8669/commits/c52a4f85f0966f7c2eeed8e267a374aff3480695.

File: src/plugins/formulas/__tests__/formulas.spec.js (hunk 12 above)
Comment (jansiegel) [reply src/plugins/formulas/__tests__/formulas.spec.js]: Rewritten this test case to check if this setup crashes the table in https://github.com/handsontable/handsontable/pull/8669/commits/c52a4f85f0966f7c2eeed8e267a374aff3480695.

**Review by wojciechczerniak — APPROVED** [review]
Thanks for addressing all my comments. Looks good now 👍
